File: 0199) Binary Tree Right Side View.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> rightSideView(TreeNode* root) {
        vector<int> ans;
        helper(root, 1, ans);
        return ans;
    }
    void helper(TreeNode *root, int i, vector<int> &ans)
    {
        if(root == nullptr)
        {
            return;
        }
        if(ans.size() < i)
        {
            ans.push_back(root->val);
        }
        helper(root->right, i+1, ans);
        helper(root->left, i+1, ans);
    }
};

File: 0199) Binary Tree Right Side View.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> ans = new ArrayList<>();
        dfs(root, ans, 1);
        return ans;
    }
    void dfs(TreeNode root, List<Integer> ans, int i)
    {
        if(root == null)
        {
            return;
        }
        if(ans.size() < i)
        {
            ans.add(root.val);
        }
        dfs(root.right, ans, i+1);
        dfs(root.left, ans, i+1);
    }
}
